File: Packages/com.satie.lang/Runtime/Core/InterpolationData.cs
using UnityEngine;

namespace Satie
{
    public enum InterpolationType
    {
        Interpolate,
        Goto,
        GoBetween
    }

    public class InterpolationData
    {
        public RangeOrValue minRange;
        public RangeOrValue maxRange;
        public RangeOrValue durationRange;
        public string easeName;
        public float minValue;
        public float maxValue;
        public int repeatCount;
        public bool isForever;
        public InterpolationType interpolationType;

        public InterpolationData(RangeOrValue min, RangeOrValue max, string easeName, RangeOrValue dur, int count = 1, bool forever = false, InterpolationType type = InterpolationType.Interpolate)
        {
            minRange = min;
            maxRange = max;
            durationRange = dur;
            this.easeName = easeName;
            interpolationType = type;
            repeatCount = count;
            isForever = forever;

            minValue = min.Sample();
            maxValue = max.Sample();
        }

        public static InterpolationData Parse(string interpolateStr)
        {
            if (string.IsNullOrWhiteSpace(interpolateStr)) return null;

            // Try goto with optional easing (defaults to linear)
            // Pattern: goto(0and1 in 2) or goto(0and1 as inquad in 2)
            var gotoPattern = @"goto\s*\(\s*(?<min>-?[\d.]+(?:to-?[\d.]+)?)\s*and\s*(?<max>-?[\d.]+(?:to-?[\d.]+)?)\s*(?:as\s+(?<ease>\w+))?\s+in\s+(?<dur>-?[\d.]+(?:to-?[\d.]+)?)\s*\)";
            var gotoRegex = new System.Text.RegularExpressions.Regex(gotoPattern, System.Text.RegularExpressions.RegexOptions.IgnoreCase);
            var gotoMatch = gotoRegex.Match(interpolateStr);

            if (gotoMatch.Success)
            {
                RangeOrValue min = RangeOrValue.Parse(gotoMatch.Groups["min"].Value);
                RangeOrValue max = RangeOrValue.Parse(gotoMatch.Groups["max"].Value);
                string easeName = gotoMatch.Groups["ease"].Success ? gotoMatch.Groups["ease"].Value : "linear";
                RangeOrValue duration = RangeOrValue.Parse(gotoMatch.Groups["dur"].Value);
                return new InterpolationData(min, max, easeName, duration, 1, false, InterpolationType.Goto);
            }

            // Try gobetween with optional easing (defaults to linear)
            // Pattern: gobetween(0and1 in 2) or gobetween(0 and 1 in 2) or gobetween(0and1 as inquad in 2)
            var goBetweenPattern = @"gobetween\s*\(\s*(?<min>-?[\d.]+(?:to-?[\d.]+)?)\s*and\s*(?<max>-?[\d.]+(?:to-?[\d.]+)?)\s*(?:as\s+(?<ease>\w+))?\s+in\s+(?<dur>-?[\d.]+(?:to-?[\d.]+)?)\s*(?:\s+for\s+(?<count>ever|\d+))?\s*\)";
            var goBetweenRegex = new System.Text.RegularExpressions.Regex(goBetweenPattern, System.Text.RegularExpressions.RegexOptions.IgnoreCase);
            var goBetweenMatch = goBetweenRegex.Match(interpolateStr);

            if (goBetweenMatch.Success)
            {
                RangeOrValue min = RangeOrValue.Parse(goBetweenMatch.Groups["min"].Value);
                RangeOrValue max = RangeOrValue.Parse(goBetweenMatch.Groups["max"].Value);
                string easeName = goBetweenMatch.Groups["ease"].Success ? goBetweenMatch.Groups["ease"].Value : "linear";
                RangeOrValue duration = RangeOrValue.Parse(goBetweenMatch.Groups["dur"].Value);

                bool forever = true;
                int count = 1;

                if (goBetweenMatch.Groups["count"].Success)
                {
                    string countStr = goBetweenMatch.Groups["count"].Value.ToLower();
                    if (countStr == "ever")
                    {
                        forever = true;
                    }
                    else
                    {
                        forever = false;
                        count = int.Parse(countStr);
                    }
                }

                return new InterpolationData(min, max, easeName, duration, count, forever, InterpolationType.GoBetween);
            }

            var pattern = @"interpolate\s*\(\s*(?<min>-?[\d.]+(?:to-?[\d.]+)?)\s*and\s*(?<max>-?[\d.]+(?:to-?[\d.]+)?)\s+as\s+(?<ease>\w+)\s+in\s+(?<dur>-?[\d.]+(?:to-?[\d.]+)?)\s*(?:\s+for\s+(?<count>ever|\d+))?\s*\)";
            var regex = new System.Text.RegularExpressions.Regex(pattern, System.Text.RegularExpressions.RegexOptions.IgnoreCase);
            var match = regex.Match(interpolateStr);

            if (!match.Success) return null;

            RangeOrValue minLegacy = RangeOrValue.Parse(match.Groups["min"].Value);
            RangeOrValue maxLegacy = RangeOrValue.Parse(match.Groups["max"].Value);
            string easeNameLegacy = match.Groups["ease"].Value;
            RangeOrValue durationLegacy = RangeOrValue.Parse(match.Groups["dur"].Value);

            bool foreverLegacy = false;
            int countLegacy = 1;

            if (match.Groups["count"].Success)
            {
                string countStr = match.Groups["count"].Value.ToLower();
                if (countStr == "ever")
                {
                    foreverLegacy = true;
                }
                else
                {
                    countLegacy = int.Parse(countStr);
                }
            }

            return new InterpolationData(minLegacy, maxLegacy, easeNameLegacy, durationLegacy, countLegacy, foreverLegacy, InterpolationType.Interpolate);
        }
    }
}

File: satie-web/src/engine/core/InterpolationData.ts
/**
 * Interpolation data for dynamic property changes.
 * Supports goto(), gobetween(), and interpolate() syntax.
 * Ported from Satie Unity: InterpolationData.cs
 */
import { RangeOrValue } from './RangeOrValue';

export enum InterpolationType {
  Interpolate = 'interpolate',
  Goto = 'goto',
  GoBetween = 'gobetween',
}

export class InterpolationData {
  minRange: RangeOrValue;
  maxRange: RangeOrValue;
  durationRange: RangeOrValue;
  easeName: string;
  minValue: number;
  maxValue: number;
  repeatCount: number;
  isForever: boolean;
  interpolationType: InterpolationType;

  constructor(
    min: RangeOrValue,
    max: RangeOrValue,
    easeName: string,
    dur: RangeOrValue,
    count: number = 1,
    forever: boolean = false,
    type: InterpolationType = InterpolationType.Interpolate,
  ) {
    this.minRange = min;
    this.maxRange = max;
    this.durationRange = dur;
    this.easeName = easeName;
    this.interpolationType = type;
    this.repeatCount = count;
    this.isForever = forever;
    this.minValue = min.sample();
    this.maxValue = max.sample();
  }

  /**
   * Parse interpolation strings:
   *   goto(0and1 in 2)
   *   goto(0and1 as inquad in 2)
   *   gobetween(0and1 in 2)
   *   gobetween(0and1 as inquad in 2 for 3)
   *   gobetween(0and1 in 2 for ever)
   *   interpolate(0.8and1.2 as incubic in 10)
   */
  static parse(str: string): InterpolationData | null {
    if (!str || !str.trim()) return null;

    // goto(min and max [as ease] in dur)
    const gotoRx = /goto\s*\(\s*(?<min>-?[\d.]+(?:to-?[\d.]+)?)\s*and\s*(?<max>-?[\d.]+(?:to-?[\d.]+)?)\s*(?:as\s+(?<ease>\w+))?\s+in\s+(?<dur>-?[\d.]+(?:to-?[\d.]+)?)\s*\)/i;
    const gotoMatch = str.match(gotoRx);
    if (gotoMatch?.groups) {
      const { min, max, ease, dur } = gotoMatch.groups;
      return new InterpolationData(
        RangeOrValue.parse(min),
        RangeOrValue.parse(max),
        ease ?? 'linear',
        RangeOrValue.parse(dur),
        1,
        false,
        InterpolationType.Goto,
      );
    }

    // gobetween(min and max [as ease] in dur [for count|ever])
    const gbRx = /gobetween\s*\(\s*(?<min>-?[\d.]+(?:to-?[\d.]+)?)\s*and\s*(?<max>-?[\d.]+(?:to-?[\d.]+)?)\s*(?:as\s+(?<ease>\w+))?\s+in\s+(?<dur>-?[\d.]+(?:to-?[\d.]+)?)\s*(?:\s+for\s+(?<count>ever|\d+))?\s*\)/i;
    const gbMatch = str.match(gbRx);
    if (gbMatch?.groups) {
      const { min, max, ease, dur, count } = gbMatch.groups;
      let forever = true;
      let repeatCount = 1;
      if (count) {
        if (count.toLowerCase() === 'ever') {
          forever = true;
        } else {
          forever = false;
          repeatCount = parseInt(count, 10);
        }
      }
      return new InterpolationData(
        RangeOrValue.parse(min),
        RangeOrValue.parse(max),
        ease ?? 'linear',
        RangeOrValue.parse(dur),
        repeatCount,
        forever,
        InterpolationType.GoBetween,
      );
    }

    // interpolate(min and max as ease in dur [for count|ever])
    const interpRx = /interpolate\s*\(\s*(?<min>-?[\d.]+(?:to-?[\d.]+)?)\s*and\s*(?<max>-?[\d.]+(?:to-?[\d.]+)?)\s+as\s+(?<ease>\w+)\s+in\s+(?<dur>-?[\d.]+(?:to-?[\d.]+)?)\s*(?:\s+for\s+(?<count>ever|\d+))?\s*\)/i;
    const interpMatch = str.match(interpRx);
    if (interpMatch?.groups) {
      const { min, max, ease, dur, count } = interpMatch.groups;
      let forever = false;
      let repeatCount = 1;
      if (count) {
        if (count.toLowerCase() === 'ever') {
          forever = true;
        } else {
          repeatCount = parseInt(count, 10);
        }
      }
      return new InterpolationData(
        RangeOrValue.parse(min),
        RangeOrValue.parse(max),
        ease,
        RangeOrValue.parse(dur),
        repeatCount,
        forever,
        InterpolationType.Interpolate,
      );
    }

    return null;
  }
}
